### PR TITLE
Added different waveform logic!

### PR DESCRIPTION
**Pull request by rajatkriplani**
`libmonotone.h` now has an enum for `waveform_type_t`,  member `waveform_type` in the `monotone_s` struct, and a function prototype `monotone_set_waveform_type()`. Still when I run it I get the error:
```
./monotone_converter "songs/Jakim - Aboriginal Derivatives.mon" output.pcm
Invalid .mon file format.
```

### CODE REVIEW

File: libmonotone.c [hunk 1]
```diff
@@ -155,23 +156,54 @@ void monotone_deinit(monotone_t* monotone) {
     monotone->tracks = NULL;
 }
 
+//function to set waveform
+void monotone_set_waveform_type(monotone_t* monotone, waveform_type_t type) {
```
Comment (aspizu) [inline libmonotone.c:160]: users can just mutate `monotone->waveform_type`, this is not object-oriented programming.
Also, `waveform_type` isn't a member on `monotone_t` objects, its a member on `monotone_track_t` objects according to your code.

File: libmonotone.c [hunk 2]
```diff
@@ -155,23 +156,54 @@ void monotone_deinit(monotone_t* monotone) {
...
 size_t monotone_generate(monotone_t* monotone, uint8_t* buffer, size_t sample_count, size_t samples_per_tick) {
     size_t samples_generated = 0;
     size_t tick_count = sample_count / samples_per_tick;
+
     for (size_t i = 0; i < tick_count; i++) {
         if (!monotone_tick(monotone)) break;
+
         for (size_t j = 0; j < samples_per_tick; j++) {
-            size_t sample = 0;
+            float sample = 0;
```
Comment (aspizu) [inline libmonotone.c:172]: I would like this to not use floating-point arithmetic.

File: libmonotone.c [hunk 3]
```diff
@@ -155,23 +156,54 @@ void monotone_deinit(monotone_t* monotone) {
...
+
             for (size_t k = 0; k < monotone->total_tracks; k++) {
                 monotone_track_t* track = &monotone->tracks[k];
                 if (track->note == NOTE_OFF) continue;
-                sample += (monotone->time * track->hz * 5 / (monotone->sample_rate * 2)) % 256 < 127 ? 255 : 0;
+
+                float t = (float)monotone->time * track->hz / monotone->sample_rate;
+
+                // Waveform of choice logic
+                switch (monotone->waveform_type) {
```
Comment (aspizu) [inline libmonotone.c:181]: Could you use if-else instead of switch case?

**Review by aspizu — CHANGES_REQUESTED** [review]
Try other songs like `songs/SWINGIN1.MON` or `songs/YESSO.MON`

**Comment by rajatkriplani** [comment]
I have made changes. I have got a question! why if-else instead of switch case?

**Comment by aspizu** [comment]
```
+ gcc -std=c11 libmonotone.c main.c -o libmonotone
libmonotone.c: In function ‘monotone_generate’:
libmonotone.c:180:55: error: ‘M_PI’ undeclared (first use in this function)
  180 |                     track_sample = (size_t)(((sin(2 * M_PI * t / 100) + 1) * 127.5));
      |                                                       ^~~~
libmonotone.c:180:55: note: each undeclared identifier is reported only once for each function it appears in
```

I've pushed a commit which fixes the invalid .mon file error. Please make sure it compiles.

**Comment by rajatkriplani** [comment]
> ```
> + gcc -std=c11 libmonotone.c main.c -o libmonotone
> libmonotone.c: In function ‘monotone_generate’:
> libmonotone.c:180:55: error: ‘M_PI’ undeclared (first use in this function)
>   180 |                     track_sample = (size_t)(((sin(2 * M_PI * t / 100) + 1) * 127.5));
>       |                                                       ^~~~
> libmonotone.c:180:55: note: each undeclared identifier is reported only once for each function it appears in
> ```
> 
> I've pushed a commit which fixes the invalid .mon file error. Please make sure it compiles.

When you compile it with `gcc -o libmonotone main.c libmonotone.c -lm` there is no error:  ‘M_PI’ undeclared
but to run it compile it with `gcc -std=c11 libmonotone.c main.c -o libmonotone` I've fixed the error: ‘M_PI’ undeclared
by adding:
```c
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif
```
but then it will show error for `sin()` function to fix that you will have to tell the linker to include the math library by `-lm`
still there's the same `invalid nom error` even after updating with:
```
git checkout master
git pull origin master
```
but then again when I run the code without any changes (code currently on the repo) it runs perfectly and both the code (libmonotone repo, and on local repo) are same except the different waveform logic still. So should I just close this PR and create a new one

**Comment by aspizu** [comment]
> So should I just close this PR and create a new one

you need to resolve conflicts after pulling from upstream (not origin)

**Comment by aspizu** [comment]
and don't use math.h for sin, it uses floating-point arithmetic

**Comment by rajatkriplani** [comment]
> and don't use math.h for sin, it uses floating-point arithmetic

I am using `static const uint8_t sine_table[256]` now instead of math.h, but then I'm getting warnings like:
```bash
libmonotone.c:142:37: note: (near initialization for ‘sine_table’)
libmonotone.c:142:41: warning: excess elements in array initializer
  142 |     129,126,123,120,117,114,111,108,105,102, 99, 96, 93, 90, 87, 84
      |                                         ^~~
libmonotone.c:142:41: note: (near initialization for ‘sine_table’)
libmonotone.c:142:46: warning: excess elements in array initializer
  142 |     129,126,123,120,117,114,111,108,105,102, 99, 96, 93, 90, 87, 84
      |                                              ^~
libmonotone.c:142:46: note: (near initialization for ‘sine_table’)
libmonotone.c:142:50: warning: excess elements in array initializer
  142 |     129,126,123,120,117,114,111,108,105,102, 99, 96, 93, 90, 87, 84
  ```

**Comment by aspizu** [comment]
Your array is defined to hold at most 256 `uint8_t` elements, and your initializer (The part inside `{ ` and `}`) has more than 256 elements.

How did you generate this look-up table? Honestly, I'm getting an impression that you used some kind of AI to generate this, but you haven't understood what it generated.

Also, update your branch to resolve merge conflicts.

**Comment by rajatkriplani** [comment]
> Your array is defined to hold at most 256 `uint8_t` elements, and your initializer (The part inside `{ ` and `}`) has more than 256 elements.
> 
> How did you generate this look-up table? Honestly, I'm getting an impression that you used some kind of AI to generate this, but you haven't understood what it generated.
> 
> Also, update your branch to resolve merge conflicts.

Well I simply googled "how to use sine without math.h" it yield some results like using the sine formula which includes factorial and division, sine table witch values of each degree, but all were resulting in floating point number. Then I came across this [sine lookup table be DeepBlueMbedded](https://deepbluembedded.com/sine-lookup-table-generator-calculator/) which is used to calculate the Amplitude of the wave form by:
```c
track->phase += (track->hz * PHASE_INCREMENT) / monotone->sample_rate;
uint8_t index = (track->phase >> 24) & 0xFF;
track_sample = sine_table[index];
sample += track_sample;
```
Can you have a look at my code and give me feedback and tell me what is actually going wrong?

**Comment by aspizu** [comment]
> Can you have a look at my code and give me feedback and tell me what is actually going wrong?
sure, msg me on discord.

**Pull request closed without merging**